### PR TITLE
Prototype for Wasm code transform

### PR DESCRIPTION
**Pull request by chriseth**
Depends on #6547

### CODE REVIEW

File: libyul/backends/wasm/WasmDialect.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+		"gt_u",
+		"le_u",
+		"ge_u"
+	})
+		addFunction(name, 2, 1);
+
+	addFunction("drop", 1, 0);
+	addFunction("store", 2, 0);
+	addFunction("load", 1, 1);
+	addFunction("unreacheable", 0, 0);
```
Comment (leonardoalt) [inline libyul/backends/wasm/WasmDialect.cpp]:
```suggestion
	addFunction("unreachable", 0, 0);
```

File: libyul/backends/wasm/WasmDialect.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+
```
Comment (leonardoalt) [inline libyul/backends/wasm/WasmDialect.cpp]: extra blank line

**Comment by chriseth** [comment]
```
> solc --strict-assembly --machine ewasm /tmp/x.yul
Warning: Yul and its optimizer are still experimental. Please use the output with care.

======= /tmp/x.yul (eWasm) =======

Pretty printed source:
object "object" {
    code {
        function main()
        {
            if eqz(f(2, 4))
            {
                drop(f(5, 6))
            }
        }
        function f(a, b) -> c
        {
            c := add(mul(a, 10), 5)
        }
    }
}


Binary representation:
No binary representation found.

Text representation:
(module

(func $main
    (if (i64.eqz (call $f (i64.const 2) (i64.const 4))) (then
        (drop (call $f (i64.const 5) (i64.const 6)))
    ))
)

(func $f
    (param $a i64)
    (param $b i64)
    (result i64)
    (local $c i64)
    (set_local $c (i64.add (i64.mul (get_local $a) (i64.const 10)) (i64.const 5)))
    (get_local $c)
)

)
```

File: libyul/CMakeLists.txt [hunk 3]
```diff
@@ -40,6 +40,12 @@ add_library(yul
 	backends/evm/NoOutputAssembly.cpp
 	backends/wasm/WasmDialect.cpp
 	backends/wasm/WasmDialect.h
+	backends/wasm/EWasmObjectCompiler.cpp
```
Comment (erak) [inline libyul/CMakeLists.txt]: Could be ordered alphabetically.

**Comment by chriseth** [comment]
@erak is that your only comment? :)

**Comment by chriseth** [comment]
Updated.

File: libyul/backends/wasm/EWasmCodeTransform.cpp [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+}
+
+wasm::Expression EWasmCodeTransform::operator()(FunctionDefinition const&)
+{
+	yulAssert(false, "Should not have visited here.");
+	return {};
+}
+
+wasm::Expression EWasmCodeTransform::operator()(ForLoop const& _for)
+{
```
Comment (sifmelcara) [inline libyul/backends/wasm/EWasmCodeTransform.cpp:205]: Probably want to generate code for `_for.pre` or assert that `_for.pre` is empty?

File: libyul/backends/wasm/EWasmToText.h [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	std::string operator()(wasm::If const& _if);
+	std::string operator()(wasm::Loop const& _loop);
+	std::string operator()(wasm::Break const& _break);
+	std::string operator()(wasm::Continue const& _continue);
+	std::string operator()(wasm::Block const& _block);
+
+private:
+	std::string indented(std::string const& _in);
+
+	std::string transform(wasm::FunctionDefinition const& _function);
```
Comment (sifmelcara) [inline libyul/backends/wasm/EWasmToText.h:57]: I don't get the idea of this naming.. Why is it called transform?

File: libyul/backends/wasm/EWasmCodeTransform.cpp [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+	yulAssert(false, "EVM instruction used for Wasm code generation.");
+	return {};
+}
+
+wasm::Expression EWasmCodeTransform::operator()(If const& _if)
+{
+	return wasm::If{visit(*_if.condition), visit(_if.body.statements), {}};
+}
+
+wasm::Expression EWasmCodeTransform::operator()(Switch const& _switch)
```
Comment (sifmelcara) [inline libyul/backends/wasm/EWasmCodeTransform.cpp:159]: FWIW, `Switch` statements probably will be transformed into `If` statements in earlier phases.

File: libyul/backends/wasm/EWasmCodeTransform.cpp [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+wasm::Expression EWasmCodeTransform::generateMultiAssignment(
+	vector<string> _variableNames,
+	unique_ptr<wasm::Expression> _firstValue
+)
+{
+	yulAssert(!_variableNames.empty(), "");
+	wasm::LocalAssignment assignment{std::move(_variableNames.front()), std::move(_firstValue)};
+
+	if (_variableNames.size() == 1)
+		return move(assignment);
```
Comment (sifmelcara) [inline libyul/backends/wasm/EWasmCodeTransform.cpp:66]: Consider to add `std::`?

File: libyul/backends/wasm/EWasmToText.cpp [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+	vector<wasm::GlobalVariableDeclaration> const& _globals,
+	vector<wasm::FunctionDefinition> const& _functions
+)
+{
+	string ret = "(module\n\n";
+	for (auto const& g: _globals)
+		ret += "    (global $" + g.variableName + " (mut i64) (i64.const 0))\n";
+	for (auto const& f: _functions)
+		ret += transform(f) + "\n";
+	return move(ret) + ")\n";
```
Comment (sifmelcara) [inline libyul/backends/wasm/EWasmToText.cpp:42]: Consider to add `std::`?

File: libyul/backends/wasm/EWasmToText.cpp [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+
+string EWasmToText::joinTransformed(vector<wasm::Expression> const& _expressions)
+{
+	string ret;
+	for (auto const& e: _expressions)
+	{
+		string t = visit(e);
+		if (!t.empty() && !ret.empty() && ret.back() != '\n')
+			ret += ' ';
+		ret += move(t);
```
Comment (sifmelcara) [inline libyul/backends/wasm/EWasmToText.cpp:157]: Consider to add `std::`?

File: libyul/backends/wasm/EWasmCodeTransform.cpp [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+			statement.type() == typeid(yul::FunctionDefinition),
+			"Expected only function definitions at the highest level."
+		);
+		functions.emplace_back(transform.translateFunction(boost::get<yul::FunctionDefinition>(statement)));
+	}
+
+	return EWasmToText{}.run(transform.m_globalVariables, functions);
+}
+
+wasm::Expression EWasmCodeTransform::generateMultiAssignment(
```
Comment (sifmelcara) [inline libyul/backends/wasm/EWasmCodeTransform.cpp:57]: Maybe simply always use global variables to return values? (Even when there is only one return value)

Not sure about the pros and cons.

File: libyul/backends/wasm/EWasmToText.cpp [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+{
+	return "(set_global $" + _assignment.variableName + " " + visit(*_assignment.value) + ")\n";
+}
+
+string EWasmToText::operator()(wasm::If const& _if)
+{
+	string text = "(if " + visit(*_if.condition) + " (then\n" + indented(joinTransformed(_if.statements)) + ")";
+	if (_if.elseStatements)
+		text += "(else\n" + indented(joinTransformed(*_if.elseStatements)) + ")";
+	return std::move(text) + ")\n";
```
Comment (sifmelcara) [inline libyul/backends/wasm/EWasmToText.cpp:90]: Unnecessary `std::move`? (At least you did not use `std::move` in other functions's return statement for `std::string`)

**Review by sifmelcara — COMMENTED** [review]
Otherwise looks good to me! I'm not familiar with wasm though, I even spent ~1hr to understand wasm's `br`, `br_if`, `block`, `loop` and labels to make sure the codegen procedure for `ForLoop` is correct...

File: libyul/backends/wasm/EWasmCodeTransform.cpp (hunk 10 above)
Comment (chriseth) [reply libyul/backends/wasm/EWasmCodeTransform.cpp:57]: I would assume that global variables are way more expensive.

File: libyul/backends/wasm/EWasmCodeTransform.cpp (hunk 4 above)
Comment (chriseth) [reply libyul/backends/wasm/EWasmCodeTransform.cpp:205]: Ah thanks a lot!

File: libyul/backends/wasm/EWasmToText.cpp (hunk 11 above)
Comment (chriseth) [reply libyul/backends/wasm/EWasmToText.cpp:90]: It is not a direct `return` it first adds `")\n"` at the end.

File: libyul/backends/wasm/EWasmToText.h (hunk 5 above)
Comment (chriseth) [reply libyul/backends/wasm/EWasmToText.h:57]: It takes code in one form and transforms it into a another one...

**Comment by chriseth** [comment]
Thanks for the review! I added the pre handling, but did not have the energy to find the correct commits to add or remove the `std::` prefixes. Especially `move` is sometimes used with `std::` prefix and sometimes without in the whole codebase.

File: libyul/backends/wasm/EWasmToText.cpp (hunk 11 above)
Comment (sifmelcara) [reply libyul/backends/wasm/EWasmToText.cpp:90]: Then should add `std::move()` in `EWasmToText::operator()(wasm::Loop const& _loop)` and `EWasmToText::operator()(wasm::Block const& _block)`'s return statement? (I just trying to point out what I observed. This issue is not important at all )

File: libyul/backends/wasm/EWasmToText.h (hunk 5 above)
Comment (sifmelcara) [reply libyul/backends/wasm/EWasmToText.h:57]: I was thinking that the name is a little bit ambiguous and collides with `std::transform`. I don't mind though.

File: libyul/backends/wasm/EWasmToText.cpp (hunk 11 above)
Comment (chriseth) [reply libyul/backends/wasm/EWasmToText.cpp:90]: Yes, you are right!

**Comment by chriseth** [comment]
Updated.

File: libyul/backends/wasm/EWasmToText.cpp (hunk 8 above)
Comment (leonardoalt) [reply libyul/backends/wasm/EWasmToText.cpp:42]: We decided to only use `std::move` if there is a conflict with some other imported `move`.

File: libyul/backends/wasm/EWasmCodeTransform.cpp [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+}
+
+wasm::Expression EWasmCodeTransform::operator()(ForLoop const& _for)
+{
+	string breakLabel = newLabel();
+	string continueLabel = newLabel();
+	m_breakContinueLabelNames.push({breakLabel, continueLabel});
+
+	// The AST is constructed in this weird way because of some strange
+	// problem with move semantics.
```
Comment (ekpyron) [inline libyul/backends/wasm/EWasmCodeTransform.cpp:211]: You're probably referring to the lack of rvalue initializer lists? Unfortunately, there's no straightforward way around that... Should we add a variadic constructor for ``BuiltinCall`` et al? I.e.
```
// shared helper
template<typename T>
void variadicEmplaceBack(std::vector<T>& _vector, T&& _element)
{
  _vector.emplace_back(std::move(_element));
}
template<typename T, typename... Tail>
void variadicEmplaceBack(std::vector<T>& _vector, T&& _element, Tail&&... tail)
{
  _vector.emplace_back(std::move(_element));
  variadicEmplaceBack(_vector, std::forward<Tail>(tail)...);
}

template<typename... Expression>
BuiltinCall(std::string _name, Expression&&... _arguments): functionName(std::move(_name))
{
  variadicEmplaceBack(std::forward<Expression>(_arguments)...);
}
```
Or just a vector-construct helper like
```
template<typename T, typename... Args>
std::vector<T> make_vector(T&& _t, Args&&... _args)
{
  std::vector<T> v;
  variadicEmplaceBack(std::forward<T>(_t), std::forward<Args>(_args)...);
  return v;
}
```
Those could later be simplified with C++17.
I remember @christianparpart and I stumbling across very similar issues in the yul optimiser before, so maybe it'd be good to build a general solution for this...

Comment (chriseth) [reply libyul/backends/wasm/EWasmCodeTransform.cpp:211]: Sounds good, but I would propose to change this PR later then.

File: libyul/backends/wasm/EWasmCodeTransform.cpp [hunk 13]
```diff
@@ -57,19 +60,19 @@ wasm::Expression EWasmCodeTransform::generateMultiAssignment(
 )
 {
 	yulAssert(!_variableNames.empty(), "");
-	wasm::LocalAssignment assignment{std::move(_variableNames.front()), std::move(_firstValue)};
+	wasm::LocalAssignment assignment{move(_variableNames.front()), std::move(_firstValue)};
```
Comment (leonardoalt) [inline libyul/backends/wasm/EWasmCodeTransform.cpp:63]: Why remove only one? ;)

**Review by leonardoalt — APPROVED** [review]

File: libyul/backends/wasm/EWasmCodeTransform.cpp (hunk 13 above)
Comment (chriseth) [reply libyul/backends/wasm/EWasmCodeTransform.cpp:63]: Because I didn't use `/eg`.

**Pull request merged**